### PR TITLE
[mlir][Ptr] Init the Ptr dialect with the `!ptr.ptr` type.

### PR DESCRIPTION
**Pull request by fabianmcg**
This patch initializes the `ptr` dialect directories and base files, adding the `!ptr.ptr` type and the `#ptr.spec<...>` data layout spec attribute.

The `!ptr.ptr` type is an opaque pointer type optionally parameterized by a memory space. This type typically represents a handle to an object in memory or target-dependent values like `nullptr`.

The implementation of the `DataLayoutTypeInterface` interface for `!ptr.ptr` was adapted from `!llvm.ptr`'s implementation. This implementation uses the `#ptr.spec<...>` attribute for defining the data layout specification.

See [[RFC] `ptr` dialect & modularizing ptr ops in the LLVM dialect](https://discourse.llvm.org/t/rfc-ptr-dialect-modularizing-ptr-ops-in-the-llvm-dialect/75142) for rationale and roadmap.

### CODE REVIEW

File: mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    pointer ::= `ptr` (`<` memory-space `>`)?
+    memory-space ::= attribute-value
+    ```
+  }];
+  let parameters = (ins OptionalParameter<"Attribute">:$memorySpace);
+  let assemblyFormat = "(`<` $memorySpace^ `>`)?";
+  let builders = [
+    TypeBuilder<(ins CArg<"Attribute", "nullptr">:$addressSpace), [{
+      return $_get($_ctxt, addressSpace);
+    }]>,
```
Comment (Dinistro) [inline mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td:60]: Shouldn't this be called "memorySpace" for this case here? IIUC, the "memory space" is not limited to being an LLVM-style address space.

File: mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+                                            addressSpace));
+    }]>
+  ];
+  let skipDefaultBuilders = 1;
+  let extraClassDeclaration = [{
+    /// Returns the default memory space.
+    Attribute getDefaultMemorySpace() const;
+
+    /// Returns the memory space as an unsigned number.
+    int64_t getAddressSpace() const;
```
Comment (Dinistro) [inline mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td:72]: Should this be a ` uint64_t`  then? What is expected to happen when the "memory space" is a non-integer attribute? Does it just return the default address space then?

File: mlir/lib/Dialect/Ptr/IR/PtrTypes.cpp [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    }
+  }
+  if (currentEntry) {
+    std::optional<uint64_t> value = extractPointerSpecValue(currentEntry, pos);
+    // If the optional `PtrDLEntryPos::Index` entry is not available, use the
+    // pointer size as the index bitwidth.
+    if (!value && pos == PtrDLEntryPos::Index)
+      value = extractPointerSpecValue(currentEntry, PtrDLEntryPos::Size);
+    bool isSizeOrIndex =
+        pos == PtrDLEntryPos::Size || pos == PtrDLEntryPos::Index;
```
Comment (Dinistro) [inline mlir/lib/Dialect/Ptr/IR/PtrTypes.cpp:52]: Nit: Hoist this and reuse the result.

File: mlir/lib/Dialect/Ptr/IR/PtrTypes.cpp [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    const auto *it =
+        llvm::find_if(oldLayout, [&](DataLayoutEntryInterface entry) {
+          if (auto type = llvm::dyn_cast_if_present<Type>(entry.getKey())) {
+            return llvm::cast<PtrType>(type).getMemorySpace() ==
+                   newType.getMemorySpace();
+          }
+          return false;
+        });
+    if (it == oldLayout.end()) {
+      llvm::find_if(oldLayout, [&](DataLayoutEntryInterface entry) {
```
Comment (Dinistro) [inline mlir/lib/Dialect/Ptr/IR/PtrTypes.cpp:88]: Nit: The result of this seems unused.

**Review by Dinistro — COMMENTED** [review]
Can there be tests added for the data layout extensions?

File: mlir/lib/Dialect/Ptr/IR/PtrTypes.cpp (hunk 4 above)
Comment (gysit) [reply mlir/lib/Dialect/Ptr/IR/PtrTypes.cpp:88]: It looks like `it` should be overwritten here? `it = llvm::find_if`?

**Comment by Dinistro** [comment]
Thanks for addressing the comments. While I would like to see this landing (I'm working on changes that involve the pointer type), I will not be able to accept this. Someone from the code team should take a look, e.g., @joker-eph or @ftynse

**Comment by joker-eph** [comment]
This will have to wait post EuroLLVM I think.

**Comment by fabianmcg** [comment]
@joker-eph what about patches like #86799 and #86792 that are useful outside of the pointer dialect?

**Comment by joker-eph** [comment]
As long as you find people for these components  to review with  high confidence these in isolation, that seems fine.

**Comment by Dinistro** [comment]
Friendly ping 🙂 

Also: @fabianmcg do you have time to resolve the conflicts?

File: mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+// Pointer type definitions
+//===----------------------------------------------------------------------===//
+
+class Ptr_Type<string name, string typeMnemonic, list<Trait> traits = []>
+    : TypeDef<Ptr_Dialect, name, traits> {
+  let mnemonic = typeMnemonic;
+}
+
+def Ptr_PtrType : Ptr_Type<"Ptr", "ptr", [
+    MemRefElementTypeInterface,
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td:39]: Probably should be added later, with tests.

File: mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    }]>,
+    TypeBuilder<(ins CArg<"unsigned">:$addressSpace), [{
+      return $_get($_ctxt, IntegerAttr::get(IntegerType::get($_ctxt, 32),
+                                            addressSpace));
+    }]>
+  ];
+  let skipDefaultBuilders = 1;
+  let extraClassDeclaration = [{
+    /// Returns the default memory space.
+    Attribute getDefaultMemorySpace() const;
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td:69]: Why is this useful as a public API?
Isn't this a property of the datalayout somehow?

File: mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+    /// Returns the memory space as an unsigned number.
+    uint64_t getAddressSpace() const;
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td:72]: That forces an API where memory spaces can always be converted to integer, can we avoid this?
(it's not used right now, so just remove?)

File: mlir/lib/Dialect/Ptr/IR/PtrTypes.cpp [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+          getPointerDataLayoutEntry(params, *this, PtrDLEntryPos::Preferred))
+    return *alignment;
+
+  return dataLayout.getTypePreferredAlignment(
+      get(getContext(), getDefaultMemorySpace()));
+}
+
+std::optional<uint64_t> mlir::ptr::extractPointerSpecValue(Attribute attr,
+                                                           PtrDLEntryPos pos) {
+  auto spec = cast<DenseIntElementsAttr>(attr);
```
Comment (joker-eph) [inline mlir/lib/Dialect/Ptr/IR/PtrTypes.cpp:152]: Why aren't we using `DenseI64ArrayAttr`?

File: mlir/test/Dialect/Ptr/layout.mlir [hunk 9]
```diff
@@ -0,0 +1,87 @@
+// RUN: mlir-opt --test-data-layout-query --split-input-file --verify-diagnostics %s | FileCheck %s
+
+module attributes { dlti.dl_spec = #dlti.dl_spec<
+  #dlti.dl_entry<!ptr.ptr, dense<[32, 32, 64]> : vector<3xi64>>,
```
Comment (joker-eph) [inline mlir/test/Dialect/Ptr/layout.mlir:4]: This is a bit obscure and fragile to me, what about a dedicated attribute?
```
#dlti.dl_entry<!ptr.ptr, #ptr.spec<size=32, abi=32, preferred=64>>,
```

File: mlir/lib/Dialect/Ptr/IR/PtrTypes.cpp [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+}
+
+uint64_t PtrType::getPreferredAlignment(const DataLayout &dataLayout,
+                                        DataLayoutEntryListRef params) const {
+  if (std::optional<uint64_t> alignment =
+          getPointerDataLayoutEntry(params, *this, PtrDLEntryPos::Preferred))
+    return *alignment;
+
+  return dataLayout.getTypePreferredAlignment(
+      get(getContext(), getDefaultMemorySpace()));
```
Comment (joker-eph) [inline mlir/lib/Dialect/Ptr/IR/PtrTypes.cpp:147]: I don't understand how that does not just infinitely recurse?

I would think we'd end up here at some point: https://github.com/llvm/llvm-project/blob/main/mlir/lib/Interfaces/DataLayoutInterfaces.cpp#L219

File: mlir/lib/Dialect/Ptr/IR/PtrTypes.cpp [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+      return emitError(loc) << "expected i64 parameters for " << key;
+
+    if (extractPointerSpecValue(values, PtrDLEntryPos::Abi) >
+        extractPointerSpecValue(values, PtrDLEntryPos::Preferred)) {
+      return emitError(loc) << "preferred alignment is expected to be at least "
+                               "as large as ABI alignment";
+    }
+  }
+  return success();
+}
```
Comment (joker-eph) [inline mlir/lib/Dialect/Ptr/IR/PtrTypes.cpp:182]: I assume most of this is copied from LLVM pointer type?

File: mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td (hunk 5 above)
Comment (fabianmcg) [reply mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td:39]: I can remove it, however, there's already a first test on `types.mlir` checking `memref<!ptr.ptr>` is valid.

File: mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td (hunk 6 above)
Comment (fabianmcg) [reply mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td:69]: It can be removed. It was mostly a shortcut for the method in the memory space interface, but right now I don't have an use case.

File: mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td (hunk 7 above)
Comment (fabianmcg) [reply mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td:72]: I'll remove it.

File: mlir/lib/Dialect/Ptr/IR/PtrTypes.cpp (hunk 10 above)
Comment (fabianmcg) [reply mlir/lib/Dialect/Ptr/IR/PtrTypes.cpp:147]: That's because `getPointerDataLayoutEntry` always returns a valid value for the default memory space.

File: mlir/lib/Dialect/Ptr/IR/PtrTypes.cpp (hunk 8 above)
Comment (fabianmcg) [reply mlir/lib/Dialect/Ptr/IR/PtrTypes.cpp:152]: Many of these were carried over from LLVM. I'll update with `#ptr.spec`.

File: mlir/lib/Dialect/Ptr/IR/PtrTypes.cpp (hunk 11 above)
Comment (fabianmcg) [reply mlir/lib/Dialect/Ptr/IR/PtrTypes.cpp:182]: Indeed.

File: mlir/test/Dialect/Ptr/layout.mlir (hunk 9 above)
Comment (fabianmcg) [reply mlir/test/Dialect/Ptr/layout.mlir:4]: I agree, I'll add the attribute.

File: mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+
+def Ptr_PtrType : Ptr_Type<"Ptr", "ptr", [
+    MemRefElementTypeInterface,
+    DeclareTypeInterfaceMethods<DataLayoutTypeInterface, [
+      "areCompatible", "getIndexBitwidth", "verifyEntries"]>
+  ]> {
+  let summary = "pointer type";
+  let description = [{
+    The `ptr` type is an opaque pointer type. This type typically represents
+    a reference to an object in memory. Pointers are optionally parameterized
```
Comment (ftynse) [inline mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td:46]: Nit: pointers and references are loaded terms. I'd rather say pointer is the address of an object in memory.

File: mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+  }];
+  let parameters = (ins OptionalParameter<"Attribute">:$memorySpace);
+  let assemblyFormat = "(`<` $memorySpace^ `>`)?";
+  let builders = [
+    TypeBuilder<(ins CArg<"Attribute", "nullptr">:$memorySpace), [{
+      return $_get($_ctxt, memorySpace);
+    }]>,
+    TypeBuilder<(ins CArg<"unsigned">:$addressSpace), [{
+      return $_get($_ctxt, IntegerAttr::get(IntegerType::get($_ctxt, 32),
+                                            addressSpace));
```
Comment (ftynse) [inline mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td:63]: Do we really want to enshrine the integer address space? I'd prefer if we kept it confined to LLVM.

File: mlir/lib/Dialect/Ptr/IR/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,16 @@
+add_mlir_dialect_library(
+  MLIRPtrDialect
+  PtrAttrs.cpp
+  PtrTypes.cpp
+  PtrDialect.cpp
+  ADDITIONAL_HEADER_DIRS
+  ${PROJECT_SOURCE_DIR}/mlir/Dialect/Pointer
```
Comment (ftynse) [inline mlir/lib/Dialect/Ptr/IR/CMakeLists.txt:7]: Is this needed?

File: mlir/lib/Dialect/Ptr/IR/CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+  DEPENDS
```
Comment (ftynse) [inline mlir/lib/Dialect/Ptr/IR/CMakeLists.txt:8]: Nit: can we have a newline before this to differentiate "blocks" of items here?

File: mlir/lib/Dialect/Ptr/IR/PtrAttrs.cpp [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+//===----------------------------------------------------------------------===//
+// SpecAttr
+//===----------------------------------------------------------------------===//
+
+LogicalResult SpecAttr::verify(function_ref<InFlightDiagnostic()> emitError,
+                               uint32_t size, uint32_t abi, uint32_t preferred,
+                               uint32_t index) {
+  if (size % kBitsInByte != 0)
+    return emitError() << "size entry must be divisible by 8";
+  else if (abi % kBitsInByte != 0)
```
Comment (ftynse) [inline mlir/lib/Dialect/Ptr/IR/PtrAttrs.cpp:30]: Nit: no need to put `else` after returning from `if`.

**Review by ftynse — COMMENTED** [review]
LGTM in principle. Were there any objections to the RFC?

File: mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td (hunk 12 above)
Comment (joker-eph) [reply mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td:46]: I think the goal was to not use "address" to keep it a bit more abstract: can we say "a handle to an object in memory"?

File: mlir/lib/Dialect/Ptr/IR/PtrTypes.cpp (hunk 10 above)
Comment (joker-eph) [reply mlir/lib/Dialect/Ptr/IR/PtrTypes.cpp]: Where is `getPointerDataLayoutEntry` coming into play?

File: mlir/lib/Dialect/Ptr/IR/CMakeLists.txt (hunk 14 above)
Comment (fabianmcg) [reply mlir/lib/Dialect/Ptr/IR/CMakeLists.txt]: I removed it.

File: mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td (hunk 13 above)
Comment (fabianmcg) [reply mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td]: I've removed it. This was mainly there because of the high prevalence of code using `IntegerAttr`s for memory spaces, specially coming from MemRefs. However, I think that's ill defined and we should always require valid memory space semantics.

**Comment by fabianmcg** [comment]
> LGTM in principle. Were there any objections to the RFC?

None really, in the end we reached consensus for opaque pointers.

File: mlir/lib/Dialect/Ptr/IR/PtrTypes.cpp (hunk 10 above)
Comment (fabianmcg) [reply mlir/lib/Dialect/Ptr/IR/PtrTypes.cpp]: I removed that method in favor of `getPointerSpec`.
The recursive call invokes the layout lookup with the default memory space,  and `getPointerSpec` always returns a value for that case, breaking the recursive cycle, however, I think we should a `llvm_unreachable` or always return the spec of the default space no matter what.

File: mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td (hunk 12 above)
Comment (ftynse) [reply mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td]: I don't remember if we reached consensus on the presence of null pointers. A null pointer is not a handle to something in memory, but it's a bikeshedding discussion we can have separately.

File: mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td (hunk 13 above)
Comment (ftynse) [reply mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td]: Integer memory spaces on memrefs are a historical artifact nobody bothered to clean up...

**Review by ftynse — APPROVED** [review]

File: mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td (hunk 12 above)
Comment (fabianmcg) [reply mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td]: We didn't really get that deep onto `nullptr`, only that they might not be represented by `0` and hence a special constant should probably exist.

This a bit wordy, but what about:
```
This type typically represents a handle to an object in memory or target-dependent values like `nullptr`
```

Comment (joker-eph) [reply mlir/include/mlir/Dialect/Ptr/IR/PtrDialect.td]: LG

**Review by joker-eph — APPROVED** [review]
